### PR TITLE
WINC-741: WICD takes more responsibility of Node configuration

### PR DESCRIPTION
**Pull request by saifshaikh48**
This PR moves the version annotation application and removal to WICD,
instead of WMCO. This allows us to reduce WMCO's responsibility when reconciling
nodes -- it no longer watches or stops any WICD managed Windows services. It also
shifts the responsibility of stopping services before bootstrapping from WMCO to WICD.
To do this, the WICD cleanup command was with partial best effort cleanup, using 
most recently created services ConfigMap  if no associated node object exists.

The PR also includes miscellaneous changes:
Adds containerd to the services ConfigMap validation e2e test.
Removes hardcoded values and consolidates Machine API namespace & user data constants.
Rehomes the desired version annotation constant to the metadata pkg.

### CODE REVIEW

**Comment by saifshaikh48** [comment]
/test aws-e2e-upgrade
/test gcp-e2e-operator

fixed a validation e2e test

**Comment by saifshaikh48** [comment]
/test aws-e2e-upgrade
/test gcp-e2e-operator

Everything passed last run except the deletion test. I've added a check before removing the version annotation, this should allow node deletion to properly complete

**Comment by saifshaikh48** [comment]
No issues in local testing. Seeing [OCPBUGS-6611 - Hybrid Overlay logfile is in use and cannot be deleted](https://issues.redhat.com/browse/OCPBUGS-6611) in CI though

File: test/e2e/validation_test.go
Comment (aravindhp) [inline test/e2e/validation_test.go:483]: So this is doing more than just what the comment says. I suggest updating the comment.

File: bundle/manifests/windows-instance-config-daemon_rbac.authorization.k8s.io_v1_clusterrole.yaml
Comment (aravindhp) [inline bundle/manifests/windows-instance-config-daemon_rbac.authorization.k8s.io_v1_clusterrole.yaml:15]: Do we need to remove this permission from WMCO?

File: pkg/daemon/cleanup/cleanup.go
Comment (aravindhp) [inline pkg/daemon/cleanup/cleanup.go:98]: Please make this a separate helper in the metadata package.

File: pkg/daemon/controller/controller.go
Comment (aravindhp) [inline pkg/daemon/controller/controller.go:252]: I would rather make `ApplyAnnotations` an internal helper and expose `metadata.ApplyDesiredVersion()` and so on.

File: pkg/nodeconfig/nodeconfig.go
Comment (aravindhp) [inline pkg/nodeconfig/nodeconfig.go:432]: This function can be moved to the metadata package and be made specific to the version annotation.

File: pkg/nodeconfig/nodeconfig.go
Comment (aravindhp) [inline pkg/nodeconfig/nodeconfig.go:197]: Do we use `applyLabelsAndAnnotations()` only for the version now? If yes, then we should move that to metadata package and make it specific to the version.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @saifshaikh48

File: bundle/manifests/windows-instance-config-daemon_rbac.authorization.k8s.io_v1_clusterrole.yaml
Comment (saifshaikh48) [reply bundle/manifests/windows-instance-config-daemon_rbac.authorization.k8s.io_v1_clusterrole.yaml:15]: Nope, WMCO still patches the node with labels (worker label, BYOH label) and other annotations (public key hash, username anno)

File: pkg/nodeconfig/nodeconfig.go
Comment (saifshaikh48) [reply pkg/nodeconfig/nodeconfig.go:197]: Yep, we also use it for [other labels/annotations WMCO applies](https://github.com/openshift/windows-machine-config-operator/blob/master/pkg/nodeconfig/nodeconfig.go#L181). But let me see if I can reuse what we have in the metadata pkg to get rid of this method

File: pkg/nodeconfig/nodeconfig.go
Comment (aravindhp) [inline pkg/nodeconfig/nodeconfig.go:197]: Adding a wrapper `metadata.ApplyVersionAnnotation()` will make the code readable when paired with `metadata.WaitForVersionAnnotation()`.

File: pkg/windows/windows.go
Comment (aravindhp) [inline pkg/windows/windows.go:437]: Shouldn't you stopping services always even if the node object does not exist like we were doing before? Consider the case where we error out during kubelet setup and should start again. At that point it is best we stop all the services kubelet depends on.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/secrets/secrets.go
Comment (saifshaikh48) [reply pkg/secrets/secrets.go:20]: Are you suggesting just moving the consolidated `secrets.MachineAPINamespace` const to the `cluster` pkg? Or not consolidating at all and moving the existing constant `controller.machineAPINamespace` to `cluster`?

File: pkg/windows/windows.go
Comment (saifshaikh48) [reply pkg/windows/windows.go:941]: I didn't feet it was a super useful wrapper honestly. The analog ConfigureWICD() does more than just call ensureServiceIsRunning().

I'm still a proponent of removing it given the function it wraps takes in a simple parameter, but if more folks want to keep it that's fine with me.

File: pkg/windows/windows.go
Comment (saifshaikh48) [reply pkg/windows/windows.go:437]: This is the issue that almost made me drop this from the story's acceptance criteria. WICD [does not support cleanup if a node object doesn't exist](https://github.com/openshift/windows-machine-config-operator/blob/master/pkg/daemon/cleanup/cleanup.go#L53) - this is part of the best-effort cleanup work that was punted from the epic.

Comment (saifshaikh48) [reply pkg/windows/windows.go:437]: One way around this is to pass the desiredVersion to the cleanup command instead of relying on a node annotation, as we do to the bootstrap command. @sebsoto likely has thoughts on this, so looping him in

File: pkg/metadata/metadata.go
Comment (saifshaikh48) [reply pkg/metadata/metadata.go:105]: I'll be consistent and pass it in. We pass the context to these metadata functions at many other call sites, just not from the nodeconfig pkg.

File: test/e2e/validation_test.go
Comment (mansikulkarni96) [inline test/e2e/validation_test.go:487]: nit: If this is a boolean check, a name like inConfigMap could be sufficient

File: pkg/windows/windows.go
Comment (mansikulkarni96) [inline pkg/windows/windows.go:429]: nit: WICD binary exists

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @saifshaikh48

File: test/e2e/validation_test.go
Comment (saifshaikh48) [reply test/e2e/validation_test.go:487]: I think I'll leave this as is for readability, since it is the expected value in the assert.

File: pkg/metadata/metadata.go
Comment (aravindhp) [reply pkg/metadata/metadata.go]: I am fine with that too.

File: pkg/windows/windows.go
Comment (aravindhp) [reply pkg/windows/windows.go:941]: Having such wrappers is a good pattern and adopted in other projects like the Linux kernel. It makes the code easier to read and understandable without having to parse what the helper is actually doing.

File: pkg/windows/windows.go
Comment (aravindhp) [reply pkg/windows/windows.go:436]: It does seem weird that WICD cannot perform cleanup if the node object is not present.

>One way around this is to pass the desiredVersion to the cleanup command instead of relying on a node annotation, as we do to the bootstrap command.

I am fine with that. Why do we even need the `desiredVersion` annotation if we are passing it to WICD?

File: controllers/controllers.go
Comment (aravindhp) [inline controllers/controllers.go:30]: Please prefix the commit header with `[cluster]`

File: pkg/daemon/controller/controller.go
Comment (aravindhp) [inline pkg/daemon/controller/controller.go:254]: This can be `ApplyVersionAnnotation`

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/daemon/cleanup/cleanup.go
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:86]: Its not clear why this is being done, please document what the full effects of the `Deconfigure`   function are in the function doc. All the high level changes to the node object and vm should be captured.

File: pkg/daemon/cleanup/cleanup.go
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:57]: In the future, please separate a "rename" change like this into a different commit, it doesn't have any impact on functionality but creates noise for the real functionality changes.

File: pkg/metadata/metadata.go
Comment (sebsoto) [inline pkg/metadata/metadata.go:113]: Any reason this can't just wait until desiredVersion annotation == version annotation?

File: pkg/metadata/metadata.go
Comment (sebsoto) [inline pkg/metadata/metadata.go:92]: does this need the value arg when it can just use version.Get() internally?
Or is there a dependency issue there?

File: pkg/metadata/metadata.go
Comment (sebsoto) [reply pkg/metadata/metadata.go:113]: Also, why is node being passed in when `Get` is being called to retrieve it? The node name should be the argument here.

File: pkg/metadata/metadata.go
Comment (sebsoto) [inline pkg/metadata/metadata.go:116]: Nodes don't have a namespace, only NamespacedName.Name needs to be set

File: pkg/nodeconfig/nodeconfig.go
Comment (sebsoto) [inline pkg/nodeconfig/nodeconfig.go:205]: Outdated comment and potentially outdated logic

File: pkg/windows/windows.go
Comment (sebsoto) [reply pkg/windows/windows.go:436]: > Why do we even need the desiredVersion annotation if we are passing it to WICD

The original thought process there is that by changing the desiredVersion annotation we can trigger an upgrade without stopping and changing the WICD service, in the case where there is no change to the WICD exe.

>  One way around this is to pass the desiredVersion to the cleanup command instead of relying on a node annotation, as we do to the bootstrap command.

This works for me. Another way is to just use the latest windows-services configmap in the absence of a node object

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/daemon/cleanup/cleanup.go
Comment (saifshaikh48) [reply pkg/daemon/cleanup/cleanup.go:67]: I've split it out

File: pkg/metadata/metadata.go
Comment (saifshaikh48) [reply pkg/metadata/metadata.go]: No strong reason for either. I've taken both suggestions, thanks

File: pkg/metadata/metadata.go
Comment (saifshaikh48) [reply pkg/metadata/metadata.go]: No dependency issue, updated

File: pkg/nodeconfig/nodeconfig.go
Comment (saifshaikh48) [reply pkg/nodeconfig/nodeconfig.go]: Adjusted the comment. The logic is still fine as we want to update the nc.node object after applying all these annotations to prevent stale resource

File: pkg/windows/windows.go
Comment (saifshaikh48) [reply pkg/windows/windows.go]: > Another way is to just use the latest windows-services configmap in the absence of a node object

I went ahead and implemented this (which is half of https://issues.redhat.com/browse/WINC-853)

File: pkg/metadata/metadata.go
Comment (saifshaikh48) [reply pkg/metadata/metadata.go]: `ApplyDesiredVersionAnnotation()` can rely on `version.Get()`, but note that we do have to pass in the value for `ApplyVersionAnnotation()` since it is used from WICD. Calling functions that do `version.Get()` from WICD results in an empty version value.

File: pkg/metadata/metadata.go
Comment (aravindhp) [inline pkg/metadata/metadata.go:123]: Checking  every 15 seconds for the version annotation seems very aggressive. And is a 10 minute timeout sufficient? I suggest we introduce specific constants for this case.

File: pkg/windows/windows.go
Comment (aravindhp) [inline pkg/windows/windows.go:477]: You are overloading this function. Either rename it to `ensureWICDArtifactsExists()` and reflect what it is doing (but overloading in this manner is not good practice) or split the calls at the call site.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/windows/windows.go
Comment (saifshaikh48) [reply pkg/windows/windows.go:941]: Understood, I reverted the func removal :+1:

File: pkg/windows/windows.go
Comment (saifshaikh48) [reply pkg/windows/windows.go:477]: I'll split the calls at the call site, thanks

File: pkg/metadata/metadata.go
Comment (saifshaikh48) [reply pkg/metadata/metadata.go:123]: In my testing I saw successful version annotation application within 2 minutes. The 10 min timeout should be fine IMO. Should I change this to poll every 30 seconds instead?

Comment (aravindhp) [reply pkg/metadata/metadata.go:123]: Given what you are seeing 10 minutes is good. And if it is 2 minutes total time I am fine with 15 min intervals also.

Comment (saifshaikh48) [reply pkg/metadata/metadata.go:123]: Looks like the WICD work has sped up configuration times significantly. Machine-backed node took less than 2 min :rocket: 
```
20:34:19 - init SSH connection 
20:35:58 - start WICD controller
20:36:36 - kubelet posts ready status
20:37:13 - instance has been fully configured as a worker node
```

File: pkg/windows/windows.go
Comment (sebsoto) [inline pkg/windows/windows.go:485]: nit: A more descriptive variable name could be helpful

**Review by sebsoto — CHANGES_REQUESTED** [review]
LGTM besides a nit, will let @aravindhp approve if his comments have been addressed

**Comment by mansikulkarni96** [comment]
LGTM

File: pkg/metadata/metadata.go
Comment (aravindhp) [reply pkg/metadata/metadata.go:122]: I expect we can make this even faster by moving the HNS network removal to WICD as part of it's clean up or a separate sub-command.

**Review by aravindhp — APPROVED** [review]

**Comment by saifshaikh48** [comment]
/test gcp-e2e-operator

Cluster spin up failure, timeout 
```
INFO Updating openshift-install configmap map with the start and end times. 
…
error: failed to get configmap
```

**Comment by saifshaikh48** [comment]
/retest-required

couple of cluster install failures and a flake “waiting for no-op daemonset to scale”

**Comment by saifshaikh48** [comment]
/retest-required

Operator install failure, retesting since there have been no changes since everything was green...

**Comment by aravindhp** [comment]
@saifshaikh48 operator did not fail to install. Please see the full [log](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1397/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-operator/1623652132034973696) for the real failure. The operator eventually installs. 
```
2023/02/09 15:02:36 failed to collect logs from e2e-wm-hnp22: error waiting for job to succeed
<snip>
2023/02/09 15:04:02 failed to change e2e-cmg6s default shell to Powershell
```
Not sure if it is a flake.

The other `aws-e2c-ccm-install` failed on cluster bring up:
```
error: unable to parse image registry.build03.ci.openshift.org/ci-op-3b73d7ws/release@sha256:4493a270dade7ccd5b08876f19001942fb1682529e69e04825c6a328c5e5b734: cannot retrieve image configuration for manifest sha256:4493a270dade7ccd5b08876f19001942fb1682529e69e04825c6a328c5e5b734: Get "https://d29zazo61hdry4.cloudfront.net/docker/registry/v2/blobs/sha256/3f/3fa721b4c403cee355f28f4dec4e544715c7b6500cde8db9183a76f3bfa24d6a/data?Expires=1675946989&Signature=T2Hdn7BCu4WSfeyw89Y3JxgwxCK2Gbk2rezDHZVzIk-c2ACtgnGHW4SHOWBwGfZrZ4p5w4i7QU0P1IK51vq3fJEqulnpvYtbM1uERAvGSs6ANYrgj-Kd9VRHGBBa0DsC4RrTRyjhiQKglaz9NyxYoIVjskexRyhSPfGQ-2DEiM4nE3PY1EqO4VnYZDYbksV11NcXeVry9sY3PoAuO6MkDenABZf4~x05ACDz88GYe9RRyfruFz1MvR9gz2Onyhy69FQenrMHaOCRiN5AX4rOtmo7CRaDufLbDGwY35ukwk~ckmcgoedDXyxtboXbvW5xdG4CiHDNR2mWH67guJBrtw__&Key-Pair-Id=K35IN2XBI3W7KZ": read tcp 10.129.28.196:36408->108.138.82.27:443: read: connection reset by peer
```
We should report this if we are seeing it consistently.

**Comment by aravindhp** [comment]
/hold

I am again [seeing](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1397/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-operator/1623716281704255488):
```
 2023/02/09 17:50:40 35m0.22467367s time is required to configure 1 nodes
    create_test.go:161: 
        	Error Trace:	/go/src/github.com/openshift/windows-machine-config-operator/test/e2e/create_test.go:161
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestWMCO/create/Creation/Machine_controller
        	Messages:   	Windows node creation failed
W0209 17:50:40.706877   14646 warnings.go:70] would violate PodSecurity "restricted:latest": host namespaces (hostNetwork=true), allowPrivilegeEscalation != false (container "print-logs-e2e-wm-jrjwn" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "print-logs-e2e-wm-jrjwn" must set securityContext.capabilities.drop=["ALL"]), seccompProfile (pod or container "print-logs-e2e-wm-jrjwn" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
2023/02/09 17:50:56 failed to collect logs from e2e-wm-jrjwn: error waiting for job to succeed: 
```
@saifshaikh48 please investigate

**Comment by saifshaikh48** [comment]
/retest-required

There seems to be an infrastructure issue in azure, machine-backed nodes are taking uncharacteristically long to get configured (across PRs). Retesting to see if the behavior persists

**Comment by saifshaikh48** [comment]
/unhold

Back to normal configuration times
`16m0.159533856s time is required to configure 1 nodes`

**Pull request merged**
